docs/architecture/ADR-002-deposit-assets.md

```diff
@@ -0,0 +1,117 @@
+
+## Alternative Approaches
+
+While the proposed implementation will use the Cosmos banking module to simultaneously transfer assets and mint uTokens, it might also be possible to use a [Liquidity Pool](https://tutorials.cosmos.network/liquidity-module/) for the non-minting portion. The capital facility would be just another account which offers trades on Asset:uAsset pools (though it would still require a separate way of minting the uTokens it offers). This option should be considered if direct use of the banking module is rejected.
+
+## Decision
+
+The Cosmos `Banking` module can be used as the basis for the required capabilities.
+
+The [BaseKeeper](https://github.com/cosmos/cosmos-sdk/blob/v0.44.0/x/bank/spec/02_keepers.md) of the Cosmos `Banking` module comes with the following capabilities:
+> The base keeper provides full-permission access: the ability to arbitrary modify any account's balance and mint or burn coins.
+
+Note that `BaseKeeper` also has functions which read the total coins of each asset type in circulation on the chain, and can also read individual account balances using its embedded `ViewKeeper`.
+
+## Detailed Design
+
+The Asset Facility will have the capability to mint and burn uTokens (but not their corresponding original asset types). It will have access to an allow-list of said asset and uToken types.
+
+The Asset Facility will possess a `Module Account` to store original assets, and that module account should be forbidden from being the recipient of transactions except those intended by the module (see the warning on [this page](https://docs.cosmos.network/master/modules/bank/).)
+
+The Asset Facility should harness the Cosmos `banking` module's `BaseKeeper` for the following capabilities
+- Read the supply (chain-wide) of a given coin (both base assets and uTokens)
+- Send coins (both base assets and uTokens) from module account to user account (and vice versa)
+- Mint uTokens
+- Burn uTokens
+- `SendKeeper`: Use BlockedAddr feature to guard against unexpected transfers to module account(s)
+- `ViewKeeper`: Read individual account balances
+
+### Basic Message Types
+
+For reference, here is the `Bank` module's built in coin transfer message as seen [here](https://docs.cosmos.network/v0.39/basics/app-anatomy.html), which is used when regular users send tokens to one another:
+```go
+// MsgSend - high level transaction of the coin module
+type MsgSend struct {
+	FromAddress sdk.AccAddress `json:"from_address" yaml:"from_address"`
+	ToAddress   sdk.AccAddress `json:"to_address" yaml:"to_address"`
+	Amount      sdk.Coins      `json:"amount" yaml:"amount"`
+}
+```
+The `sdk.Coins` type is a slice (ordered list) of `sdk.Coin` which contain a token type and amount.
+
+To implement the deposit functionality of the Asset Facility, the two common message types will be:
+```go
+// MsgDeposit - a user wishes to deposit assets and receive uAssets
+type MsgDeposit struct {
```

Comment:
MsgDepositAsset sounds better - especially because there may later be something like MsgDepositCollateral later which works only with uAssets (where this one forbids uAssets and uses Assets).

Reply:
ACK